### PR TITLE
api: addition of parameter for custom filename

### LINKED ISSUE #29: RFC: Allowing custom filename requests to fallback to "default" quality
A major use case with IIIF for us is to allow the users to download archived images in different resolutions, and in JPG format. We are currently doing this, but the naming convention of IIIF poses a big problem in that all files are then called `default.jpg`.

I propose that we introduce a config variable (e.g. `IIIF_DEFAULT_FALLBACK`) to allow for custom filenames to fallback to the default quality, which is most often in use. 

I have asked this question on the [IIIF discussion group](https://groups.google.com/forum/#!topic/iiif-discuss/VkRG6-26Rzk), who seems to approve. There is also no strict requirement for not doing as I suggest (http://iiif.io/api/image/2.1/#quality).

I have also considered the HTML5 tag `download="filename.jpg"`, but the browser support for this is not very good (neither IE nor Safari supports it). Another suggested method is to utilize the Content-Disposition header, but I can't see how this can be done in a practical manner.

### PR DESCRIPTION
**Pull request by blixhavn**
* Adds checking of `dl` request parameter for providing a custom
  filename. (closes #29)

* Removes duplicate lines that triggered iiif_after_process_request
  signal twice.

Signed-off-by: Øystein Blixhavn <oystein@blixhavn.no>

### CODE REVIEW

File: flask_iiif/restful.py [hunk 1]
```diff
@@ -183,12 +184,13 @@ def get(self, version, uuid, region, size, rotation, quality,
...
-            image=to_serve
-        )
+        dlparam = secure_filename(request.args.get("dl", ''))
+        if dlparam:
+            filename = dlparam + "." + image_format
+        else:
+            filename = quality + "." + image_format
 
-        # Trigger event after proccess the api request
-        iiif_after_process_request.send(self, **api_after_request_parameters)
```
Comment (jirikuncar) [inline flask_iiif/restful.py]: @blixhavn why did you remove this signal?

Comment (blixhavn) [inline flask_iiif/restful.py]: > Removes duplicate lines that triggered iiif_after_process_request
signal twice.

File: flask_iiif/restful.py [hunk 2]
```diff
@@ -183,12 +184,13 @@ def get(self, version, uuid, region, size, rotation, quality,
         # Trigger event after proccess the api request
         iiif_after_process_request.send(self, **api_after_request_parameters)
 
-        # Built the after request parameters
-        api_after_request_parameters = dict(
-            mimetype=mimetype,
-            image=to_serve
-        )
+        dlparam = secure_filename(request.args.get("dl", ''))
```
Comment (jirikuncar) [inline flask_iiif/restful.py]: My understanding is that we should use directly the filename as specified in `dl` argument.

```python
send_file_kwargs = {'mimetype': mimetype}
filename = secure_filename(request.args.get('dl'))
if filename:
    send_file_kwargs['as_attachement'] = True
    send_file_kwargs['attachment_filename'] = filename

return send_file(to_serve, **send_file_kwargs)
```

File: flask_iiif/restful.py [hunk 3]
```diff
@@ -183,12 +184,14 @@ def get(self, version, uuid, region, size, rotation, quality,
...
-        api_after_request_parameters = dict(
-            mimetype=mimetype,
-            image=to_serve
-        )
 
-        # Trigger event after proccess the api request
-        iiif_after_process_request.send(self, **api_after_request_parameters)
-        return send_file(to_serve, mimetype=mimetype)
+        send_file_kwargs = {'mimetype': mimetype}
+        dlparam = secure_filename(request.args.get('dl', ''))
```
Comment (jirikuncar) [inline flask_iiif/restful.py]:
```python
basename = ...
if basename:
    ...
    attachment_filename='{0}.{1}'.format(basename, image_format),
```

File: flask_iiif/restful.py [hunk 4]
```diff
@@ -11,13 +11,14 @@
 
 from io import BytesIO
 
-from flask import current_app, jsonify, redirect, send_file, url_for
+from flask import current_app, request, jsonify, redirect, send_file, url_for
```
Comment (jirikuncar) [inline flask_iiif/restful.py]: `isort`

**Comment by jirikuncar** [comment]
@blixhavn please have a 👀

**Comment by blixhavn** [comment]
Looks good :)

File: flask_iiif/restful.py [hunk 5]
```diff
@@ -171,12 +172,12 @@ def get(self, version, uuid, region, size, rotation, quality,
...
-        api_after_request_parameters = dict(
-            mimetype=mimetype,
-            image=to_serve
-        )
+        send_file_kwargs = {'mimetype': mimetype}
+        basename = secure_filename(request.args.get('dl', ''))
+        if basename:
+            send_file_kwargs.update(
+                as_attachment=True,
+                attachment_filename='{0}.{1}'.format(basename, image_format),
```
Comment (drjova) [inline flask_iiif/restful.py]: hmm can we do this more generic as @greut suggested on https://github.com/inveniosoftware/flask-iiif/issues/29#issuecomment-248363833? i.e. to have an extra ``custom_filename`` for custom filenames?

Comment (greut) [inline flask_iiif/restful.py]: this code does the extra `custom_filename`. However, it doesn't do the reasonable default filename when no custom filename is provided.

Comment (drjova) [inline flask_iiif/restful.py]: yep

Comment (jirikuncar) [inline flask_iiif/restful.py]: so basically we should change the `if basename:` to `if 'dl' in request.args:`.

File: flask_iiif/restful.py [hunk 6]
```diff
@@ -183,12 +172,18 @@ def get(self, version, uuid, region, size, rotation, quality,
...
+        send_file_kwargs = {'mimetype': mimetype}
+        if 'dl' in request.args:
+            basename = secure_filename(request.args.get('dl', ''))
+            if basename.lower() in {'', '1', 'true'}:
+                basename = uuid
 
-        # Trigger event after proccess the api request
-        iiif_after_process_request.send(self, **api_after_request_parameters)
-        return send_file(to_serve, mimetype=mimetype)
+            attachement_filename = '{0}-{1}-{2}-{3}-{4}.{5}'.format(
```
Comment (jirikuncar) [inline flask_iiif/restful.py]: @drjova @blixhavn please check the new default name

Comment (greut) [inline flask_iiif/restful.py]: _[attachment](https://en.wiktionary.org/wiki/attachment)_

Comment (drjova) [inline flask_iiif/restful.py]: WDYT?

If the `dl` in ['', '1', 'true'] then return our default filename format otherwise let the user choose the filename e.x. ``?dl=foo_1.jpg`` or we can move this to a separated parameter ``custom_filename`` e.x. ``?dl=1&custom_filename=foo_1.jpg`` as @greut suggested

```python
if 'dl' in request.args:
    basename = secure_filename(request.args.get('dl', ''))
    
    if basename.lower() in {'', '1', 'true'}:
        # If it doesn't contain a specific filename just return the default  filename format
        attachment_filename = '{0}-{1}-{2}-{3}-{4}.{5}'.format(
           uuid, region, size, quality, rotation, image_format
         )
     else:
         # return the requested filename
         attachment_filename = basename

    send_file_kwargs.update(
        as_attachment=True,
        attachment_filename=secure_filename(attachment_filename),
    )



```
